packages/promise-tracker/test/tracker.test.ts

```diff
@@ -0,0 +1,135 @@
+      const target = new TrackedObject();
+
+      await target.trackedMethod();
+
+      let promises = getOutstandingPromises({ target });
+
+      assert.lengthOf(promises, 0);
+    });
+
+    it("doesn't track untracked async methods", async () => {
+      const target = new TrackedObject();
+
+      const p = target.untrackedMethod();
+
+      let promises = getOutstandingPromises({ target });
+      assert.lengthOf(promises, 0);
+
+      await p;
+    });
+  });
+
+  describe("waitForOutstandingPromises", () => {
+    it("waits on a single promise from a single instance", async () => {
+      const target = new TrackedObject();
+
+      target.trackedMethod();
+
+      assert.strictEqual(target.executedTrackedMethodCount, 0);
+      await waitForOutstandingPromises();
+      assert.strictEqual(target.executedTrackedMethodCount, 1);
+    });
+
+    it("waits on multiple promises from a single instance", async () => {
+      const target = new TrackedObject();
+
+      target.trackedMethod();
+      target.trackedMethod();
+
+      assert.strictEqual(target.executedTrackedMethodCount, 0);
+      await waitForOutstandingPromises();
+      assert.strictEqual(target.executedTrackedMethodCount, 2);
+    });
+
+    it("waits on promises for a specific instance", async () => {
+      const target1 = new TrackedObject();
+      const target2 = new TrackedObject();
+
+      target1.trackedMethod();
+      const p = target2.trackedMethod(50);
+
+      assert.strictEqual(target1.executedTrackedMethodCount, 0);
+      assert.strictEqual(target2.executedTrackedMethodCount, 0);
+      await waitForOutstandingPromises({ target: target1 });
+      assert.strictEqual(target1.executedTrackedMethodCount, 1);
+      assert.strictEqual(target2.executedTrackedMethodCount, 0);
+      await p;
+      assert.strictEqual(target2.executedTrackedMethodCount, 1);
+    });
+
+    it("waits for all promises, even when some throw", async () => {
+      const target = new TrackedObject();
+
+      // delay for longer to ensure this finishes after the one that throws
+      target.trackedMethod(50);
+
+      // throws after 25ms
+      target.throwingTrackedMethod();
+
+      assert.strictEqual(target.executedTrackedMethodCount, 0);
+      assert.strictEqual(target.executedThrowingTrackedMethodCount, 0);
+      await waitForOutstandingPromises();
+      assert.strictEqual(target.executedTrackedMethodCount, 1);
+      assert.strictEqual(target.executedThrowingTrackedMethodCount, 1);
+    });
+
+    it("allows caller to pass in custom catch handler logic", async () => {
+      const target = new TrackedObject();
+
+      let throwCount = 0;
+
+      // throws after 25ms
+      target.throwingTrackedMethod();
+      target.throwingTrackedMethod();
+
+      assert.strictEqual(throwCount, 0);
+      await waitForOutstandingPromises({ catchHandler: () => throwCount++ });
+      assert.strictEqual(throwCount, 2);
+    });
+  });
+});
+
+class TrackedObject {
```

Comment:
>...classes must be defined before they can be constructed.

TrackerObject is referenced, not constructed, before the class definition. By the time, its constructor is invoked, TrackerObject has already defined.

My mental model is (open to correction @davidmurdoch!)... Javascript runs to completion, so  it will first process the module, building up references, where the `it` function will have closed over the TrackerObject class definition.  When the `it` function runs, all should be good.

Reply:
It's technically fine here... but i think it is super weird to put helpers at the bottom.

Reply:
It runs just fine (build would be red, otherwise).

As a style choice, I tend to put helpers at the bottom to aid in human readability, as when you're reading through that file, the helper class is unlikely to be the thing you're looking for first. IIRC, @gnidan has the same preference, but I could've misinterpreted.

I think it's fair to stick this in the "style choice" bucket, as I'm not getting any `ReferenceError`s, etc, when running the code here. Personally in that case I don't really care much about which particular style we go with here. I just want it to be consistent with the rest of the codebase. With that in mind, if this is enough of an issue to request a change, I'd prefer that we encode and enforce these sort of style rules as prettier or eslint config.

Reply:
I don't really care that much (though i'd like to take this moment to point out that imports are basically helpers and so now you've got them at the top _and_ the bottom 😁 :troll: )

Reply:
I mean, if you squint hard enough, all code is helper code! 😆

Reply:
🤣